### PR TITLE
Data subsystem

### PR DESCRIPTION
**Pull request by Whathecode**
Added serializable types to represent data which will be useful to both the data subsystem backend and client. From [the new carp.data README](https://github.com/cph-cachet/carp.core-kotlin/blob/data-subsystem/docs/carp-data.md):

> `Measurement`'s contain measured `Data` at a specific point or interval in time. `Measurement`'s of the same `DataType` coming from the same device in a study deployment are aggregated in data streams, identified by `DataStreamId`.
> 
> Data streams are made up out of `DataStreamPoint`'s. Multiple subsequent `DataStreamPoint`'s can be stored as `DataStreamSequence`. To aggregate multiple `DataStreamSequence`'s of one or more data streams, e.g., to prepare for data upload, `DataStreamBatch` can be used.

A `DataStreamSequence` specifies a single `SyncPoint` and `triggerIds` for all data points contained within the sequence. Application services use the readonly `DataStreamSequence` and `DataStreamBatch` interfaces. Mutable counterparts (`MutableDataStreamSequence` and `MutableDataStreamBatch`) are available to initialize these. Serializers are provided which operate on the interfaces. I believe this is very similar to how Kotlin implemented `List`.

Down the line, given that `DataStreamBatch` is a sequence, I expect that we can implement test data generators through `DataStreamBatch` to mock incoming sensor data.

Data stream object model

Next, I added `DataStreamService` and `DataStreamServiceHost` which exposes the `appendToDataStreams` and `getDataStream` endpoints. This service does not yet communicate in any way with other services. In a next PR after I get back from vacation I can use the event bus to verify whether the study deployment exists and whether it is still active, which later will also verify whether everyone in the deployment has given consent.

In addition, as part of these changes:
- [`EnumObjectList` has now been turned into a map](https://github.com/cph-cachet/carp.core-kotlin/pull/288/commits/98edf6619a162e1febfaf445c34c1f2e69f40131), which seemed to be the common requirement wherever this class was extended.
- `RangeExtensions` has been added: for now supporting range intersections.

### CODE REVIEW

File: carp.data.core/src/commonMain/kotlin/dk/cachet/carp/data/application/DataStreamServiceHost.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package dk.cachet.carp.data.application
+
+import dk.cachet.carp.common.application.UUID
+import dk.cachet.carp.data.domain.DataStreamRepository
+
+
+/**
+ * Store and retrieve [DataStreamPoint]s for study deployments.
+ */
+class DataStreamServiceHost( private val repository: DataStreamRepository ) : DataStreamService
```
Comment (ltj) [inline carp.data.core/src/commonMain/kotlin/dk/cachet/carp/data/application/DataStreamServiceHost.kt:10]: Not sure if this is to be regarded as an edge service or an internal service receiving data through different channels. If this is an edge service is should probably not have a repository, but something more flexible like a "sink" interface that can have any implementation required by the underlying infrastructure.

**Review by ltj — COMMENTED** [review]
This looks very much aligned with my current implementation. Let's have a meeting to make sure we're on the same page before proceeding.

File: carp.data.core/src/commonMain/kotlin/dk/cachet/carp/data/application/DataStreamSequence.kt [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+/**
+ * A sequence of consecutive [measurements] for a [dataStream] starting from [firstSequenceId]
+ * which all share the same [triggerIds] and [syncPoint].
+ */
+sealed interface DataStreamSequence : Sequence<DataStreamPoint<*>>
+{
+    val dataStream: DataStreamId
+    val firstSequenceId: Long
+    val measurements: List<Measurement<*>>
```
[the remark this answers is not in the document]
Comment (Whathecode) [reply carp.data.core/src/commonMain/kotlin/dk/cachet/carp/data/application/DataStreamSequence.kt:18]: My thoughts exactly, which is why I created this issue: https://github.com/cph-cachet/carp.core-kotlin/issues/289

File: carp.data.core/src/commonMain/kotlin/dk/cachet/carp/data/application/DataStreamId.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import dk.cachet.carp.common.application.data.DataType
+import kotlinx.serialization.Serializable
+
+
+/**
+ * Identifies a data stream of collected [dataType] data on the device with [deviceRoleName]
+ * in a deployed study protocol with [studyDeploymentId].
+ */
+@Serializable
+data class DataStreamId( val studyDeploymentId: UUID, val deviceRoleName: String, val dataType: DataType )
```
[the remark this answers is not in the document]
Comment (Whathecode) [reply carp.data.core/src/commonMain/kotlin/dk/cachet/carp/data/application/DataStreamId.kt:13]: I looked at the mobile core codebase before rolling my own. :) Some things I adopted; others I changed.

**Pull request merged**